### PR TITLE
Add MultiCategoricalProjectionNetwork

### LINKED ISSUE #694: Projection Network for more than 1 action with differing action spaces
Following the discussion from #37, I developed a `MultiCategoricalProjectionNetwork` that splits logits and creates the respective `Categorical` distribution. I tried to adhere to the same pattern as far as I could; It can be found here: https://gist.github.com/sidney-tio/66abada949f1b629dd9ee28777d402d5

If the team would like, I could raise a PR based on the gist I developed. From what I see, these are the to-dos to make it PR-worthy:

- [x] add tests 
- [x] add more detailed docstrings
- [x] add masks

### PR DESCRIPTION
**Pull request by sidney-tio**
Closes #694

### CODE REVIEW

**Comment by sibyjackgrove** [comment]
@sidney-tio Thank you for this PR. It could also solve #702 for me. Hope that this will be merged soon.

**Comment by sidney-tio** [comment]
hi team! could i request a review for this please?

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+  def _create_distrib(self, logits):
+    logits_splitted = tf.split(logits, self.categories_shape, -1)
+    distribs = tf.nest.map_structure(
+                    lambda l: tfp.distributions.Categorical(logits = l),
+                    logits_splitted)
+    return distribs
+
+  def _mode(self):
+    return self._flatten_and_concat_event(
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: Why mode needs flatten_and_concat_event but sample doesn't?

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+from tf_agents.networks import network
+from tf_agents.networks import utils
+from tf_agents.specs import tensor_spec
+from tf_agents.specs import distribution_spec
+
+class MultiCategoricalDistributionBlock(tfp.distributions.Blockwise):
+
+  def __init__(self, logits, categories_shape):
+    self.categories_shape = categories_shape
+    distribs = self._create_distrib(logits)
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: Can you validate that the dimensions of logits align with categories_shape?

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  def __init__(self,
+            sample_spec,
+            logits_init_output_factor=0.1,
+            name='MultiCategoricalProjectionNetwork'):
+    """Creates an instance of MultiCategoricalProjectionNetwork
+
+    Args:
+      sample_spec: A collection of `tensor_spec.BoundedTensorSpec` detailing
+        the shape and dtypes of samples pulled from the output distribution.
+      logits_init_output_factor: Output factor for initializing kernal
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: Not clear what does it mean output factor?
kernal -> kernel

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      name=name
+    )
+
+    self._sample_spec = sample_spec
+    self.categories_shape = categories_shape
+    self.n_unique_categories = n_unique_categories
+
+    self._projection_layer = tf.keras.layers.Dense(
+        self.n_unique_categories,
+        kernel_initializer=tf.compat.v1.keras.initializers.VarianceScaling(
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: Why using VarianceScaling as initializer? Can you use just tf.keras.initializers?

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            scale=logits_init_output_factor),
+        bias_initializer=tf.keras.initializers.Zeros(),
+        name='logits'
+        )
+
+  def _categories_shape(self, sample_spec):
+    def _get_n_categories(array_spec):
+      if not tensor_spec.is_bounded(array_spec):
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: Simplify to
if tensor_spec.is_bounded(array_spec) and tensor_spec.is_discrete(array_spec) :
  n_categories = array_spec.maximum - array_spec.minimum + 1
  return n_categories
else:
  raise ValueError('sample_spec must be discrete and bounded. Got: %s.' % array_spec)

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+      if not tensor_spec.is_discrete(array_spec):
+        raise ValueError('sample_spec must be discrete. Got: %s.' % array_spec)
+
+      n_categories = array_spec.maximum - array_spec.minimum +1
+      return n_categories
+
+    flattened_spec = tf.nest.flatten(sample_spec)
+    categories_shape = tf.nest.map_structure(
+      lambda s: _get_n_categories(s),
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: lambda s: _get_n_categories(s) -> _get_n_categories

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            )
+    }
+    return distribution_spec.DistributionSpec(
+        MultiCategoricalDistributionBlock,
+        input_param_spec,
+        sample_spec=sample_spec,
+        categories_shape = categories_shape
+        )
+
+  def call(self, inputs, outer_rank, training=False, mask=None):
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: Does outer_rank changes between calls? If it is fixed then better to move to __init__

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    inputs = batch_squash.flatten(inputs)
+    inputs = tf.cast(inputs, tf.float32)
+
+    logits = self._projection_layer(inputs, training=training)
+    logits = tf.reshape(logits, [-1] + [self.n_unique_categories])
+    logits = batch_squash.unflatten(logits)
+    if mask is not None:
+      # assume mask is a flattened array for now
+
+      # If the action spec says each action should be shaped (1,), add another
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: Keep lines under 80 characters.

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  def call(self, inputs, outer_rank, training=False, mask=None):
+    batch_squash = utils.BatchSquash(outer_rank)
+    inputs = batch_squash.flatten(inputs)
+    inputs = tf.cast(inputs, tf.float32)
+
+    logits = self._projection_layer(inputs, training=training)
+    logits = tf.reshape(logits, [-1] + [self.n_unique_categories])
+    logits = batch_squash.unflatten(logits)
+    if mask is not None:
+      # assume mask is a flattened array for now
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: I suppose mask should have the same shape as actions.

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+      # dimension so the final shape is (B, 1, A), where A is the number of
+      # actions. This will make Categorical emit events shaped (B, 1) rather
+      # than (B,). Using axis -2 to allow for (B, T, 1, A) shaped q_values.
+      if mask.shape.rank < logits.shape.rank:
+        mask = tf.expand_dims(mask, -2)
+
+      # Overwrite the logits for invalid actions to a very large negative
+      # number. We do not use -inf because it produces NaNs in many tfp
+      # functions.
+      almost_neg_inf = tf.constant(logits.dtype.min, dtype = logits.dtype)
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: Make sure lines are <80 and parameters don't have extra spaces.
almost_neg_inf = tf.constant(logits.dtype.min, 
    dtype=logits.dtype)

File: tf_agents/networks/multi_categorical_projection_network.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      logits = tf.compat.v2.where(
+        tf.cast(mask, tf.bool), logits, almost_neg_inf)
+
+    return self.output_spec.build_distribution(logits= logits), ()
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network.py]: (logits= logits) -> (logits=logits)

File: tf_agents/networks/multi_categorical_projection_network_test.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+def _get_inputs(batch_size, num_input_dims):
+  return tf.random.uniform([batch_size, num_input_dims])
+
+class MultiCategoricalProjectionNetworkTest(tf.test.TestCase):
+
+  def testBuild(self):
+    output_spec = [
+      tensor_spec.BoundedTensorSpec([], tf.int32, 0, 1),
+      tensor_spec.BoundedTensorSpec([], tf.int32, 0, 4)]
+    network = multi_categorical_projection_network.MultiCategoricalProjectionNetwork(
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network_test.py]: too long of a name, can you make it shorter multi_categorical_network.MultiCategoricalNetwork

File: tf_agents/networks/multi_categorical_projection_network_test.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    distribution, _ = network(inputs, outer_rank=1)
+    self.evaluate(tf.compat.v1.global_variables_initializer())
+    sample = self.evaluate(distribution.sample())
+
+    self.assertEqual(multi_categorical_projection_network.MultiCategoricalDistributionBlock,
+      type(distribution))
+    self.assertEqual(tfp.distributions.Categorical, type(distribution.distributions[0]))
+    self.assertEqual(2, len(distribution.distributions))
+    self.assertEqual((3, 7), distribution._parameters['logits'].shape)
+    self.assertEqual((3, 2), sample.shape)
```
Comment (sguada) [inline tf_agents/networks/multi_categorical_projection_network_test.py]: Can you also test that the samples respect the bounds?

**Review by sguada — CHANGES_REQUESTED** [review]

File: tf_agents/networks/multi_categorical_projection_network.py (hunk 1 above)
Comment (sidney-tio) [reply tf_agents/networks/multi_categorical_projection_network.py]: `Blockwise` from tf-probability doesn't implement `mode`, but implements `flatten_and_concat_event` for `sample` and `mean`. 

https://github.com/tensorflow/probability/blob/88d217dfe8be49050362eb14ba3076c0dc0f1ba6/tensorflow_probability/python/distributions/blockwise.py#L347-L366

File: tf_agents/networks/multi_categorical_projection_network.py (hunk 3 above)
Comment (sidney-tio) [reply tf_agents/networks/multi_categorical_projection_network.py]: Fixed kernel spelling error. 

The descriptor for `logits_init_output_factor` was inherited from `CategoricalProjectionNetwork`, but I do agree it is not exactly clear what output factor means here.

File: tf_agents/networks/multi_categorical_projection_network.py (hunk 4 above)
Comment (sidney-tio) [reply tf_agents/networks/multi_categorical_projection_network.py]: Similar to output factor, the initializer was selected for `CategoricalProjectionNetwork`, to which I'm not sure if there's any reason behind VarianceScaling used here

https://github.com/tensorflow/agents/blob/53606857476f0ac638da4a20e5eb973f4ebc1ef6/tf_agents/networks/categorical_projection_network.py#L81-L82

File: tf_agents/networks/multi_categorical_projection_network.py (hunk 9 above)
Comment (sidney-tio) [reply tf_agents/networks/multi_categorical_projection_network.py]: I'm not sure about the more appropriate approach for this so any advice here would be good; At this stage the outputs of the network exists as logits, i.e. the vector exists as flattened. In this case, `mask` needs to be flattened as well. 

Alternatively, if we want the `mask` to be same shape as actions, we could do the masking during __init__ of `MultiCategoricalDistributionBlock`

**Comment by sidney-tio** [comment]
thanks for the review! made the changes as advised. Some of the changes requested were for code inherited from CategoricalProjectionNetwork, so I'm not sure if it warrants a separate PR to reconcile both?

**Comment by MiriamJo** [comment]
@sguada Is there a reason this was never merged? I think this one would make such a difference.

**Comment by tfboyd** [comment]
Closing, after this much time I do not think anyone will get to it.  I am merging others that I can review myself. This one is too big for me. Sorry this one got so far and didn't land. :-(

**Comment by tfboyd** [comment]
@sguada  just for visibility.
